docs/proposals/image-spec.md

```diff
@@ -0,0 +1,480 @@
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+## Abstract
+
+A proposal for adding an `ImageSpec` type to container specs, to facilitate alternative image formats and more
+deterministic image references.
+
+## Motivation
+
+Currently, all of the information about where to locate an image or how to verify the image is embedded in a single
+string in the container spec. However that opaque string is not structured and sometimes confusing for users if they
+are not familiar with how the string is being parsed.
+
+For example, `gcr.io/google_containers/nginx` refers to an image located in `gcr.io` registry, with the repository path
+as `google_containers/etcd`; While `library/nginx` refers to an image located in `docker.io` registry, with the repository path
+as `library/nginx`, which is same as `nginx` because it is an [official Docker repository](https://docs.docker.com/docker-hub/official_repos/).
+
+Besides, having only one string to represent the image precludes the possibility of adding any other image types
+such as [ACI](https://github.com/appc/spec), [Fabric8 Java containers](http://fabric8.io/gitbook/javaContainer.html),
+etc.
+
+The goal of this proposal is to add an *optional* `ImageSpec` field in the `Container` object,
+which will give users the ability to:
+
+1. Express the image types they want to use.
+2. Express how and where they want the images to be fetched from.
+
+## Design
+
+1. Add an optional field called `ImageSpec` in the `Container` object and make `Image` field optional.
+2. For GET requests, We will show both the `Image` and `ImageSpec` fields for the sake of backward compatibility.
+3. For other requests, either `Image` or `ImageSpec` should be specified but not both. These two fields are
+   mutually exclusive. However `ImageSpec` will always be populated internally if `Image` is being used.
+4. If `ImageSpec` is not empty, it MUST have one and ONLY one non-empty field to specify the details of the image,
+similiar to the `VolumeSource` type.
+
+#### Open Qustions
+1. Should we be able to mix different image format in one pod?
+2. Can we just use merely `ImageSpec` in the internal API objects now? However this will make the `Image` string
+   be filled with registry info when returned to users, e.g. `Image: nginx` would become `Image: docker.io/nginx`.
+   If this is unacceptable for now, then we can do it later.
+3. Do we need to also add `ImageSpec` in the `ContainerStatus`? Maybe we should just remove the `Image` field in
+   `ContainerStatus` in the future as this field cannot vary from the spec, so only `ImageSpec` in the `Container`
+   should be enough.
+
+Following is an example of how new internal API objects will look like:
+
+```go
+package api
+
+type Container struct {
+  // Other fields ommited for brevity.
+
+  // Optional, Image specifies the name of the image, it assumes Docker
+  // image format. Either Image or ImageSpec should be specified, but not both.
+  Image *string `json:"image",omitempty`
```

Comment:
Image is already optional (tagged "omitempty"). It doesn't need to be a pointer, since empty string can be used to indicate that it's not specified.